subprojects/language-java/src/integTest/groovy/org/gradle/api/tasks/JavaToolchainBuildOperationsIntegrationTest.groovy

```diff
@@ -187,67 +187,280 @@ class JavaToolchainBuildOperationsIntegrationTest extends AbstractIntegrationSpe
         assertToolchainUsages(events1, jdkMetadata1, "JavaLauncher")
         assertToolchainUsages(events2, jdkMetadata2, "JavaLauncher")
 
-        // TODO: should it work with up-to-date tasks?
-//        when:
-//        runWithInstallationPaths(installationPaths, task)
-//        events = eventsFor(taskPath)
-//        events1 = filterByJavaVersion(events, jdkMetadata1)
-//        events2 = filterByJavaVersion(events, jdkMetadata2)
-//        then:
-//        skipped(taskPath)
-//        events.size() > 0
-//        assertToolchainUsages(events1, jdkMetadata1, "JavaLauncher")
-//        assertToolchainUsages(events2, jdkMetadata2, "JavaLauncher")
+        when:
+        runWithInstallationPaths(installationPaths, ":myToolchainTask1", ":myToolchainTask2")
+        events1 = toolchainEvents(":myToolchainTask1")
+        events2 = toolchainEvents(":myToolchainTask2")
+        then:
+        skipped(":myToolchainTask1", ":myToolchainTask2")
+        assertToolchainUsages(events1, jdkMetadata1, "JavaLauncher")
+        assertToolchainUsages(events2, jdkMetadata2, "JavaLauncher")
+    }
+
+    def "emits toolchain usages for compilation that configures java home via fork options"() {
+        JvmInstallationMetadata jdkMetadata = AvailableJavaHomes.getJvmInstallationMetadata(AvailableJavaHomes.differentJdk)
+
+        buildFile << """
+            compileJava {
+                options.forkOptions.javaHome = file("${jdkMetadata.javaHome}")
+            }
+        """
+
+        file("src/main/java/Foo.java") << """
+            public class Foo { }
+        """
+
+        def task = ":compileJava"
+
+        when:
+        runWithInstallation(jdkMetadata, task)
+        def events = toolchainEvents(task)
+        then:
+        executedAndNotSkipped(task)
+        assertToolchainUsages(events, jdkMetadata, "JavaCompiler")
+
+        when:
+        runWithInstallation(jdkMetadata, task)
+        events = toolchainEvents(task)
+        then:
+        skipped(task)
+        assertToolchainUsages(events, jdkMetadata, "JavaCompiler")
+    }
+
+    def "emits toolchain usages for compilation that configures java home via fork options pointing outside installations"() {
+        JvmInstallationMetadata jdkMetadata1 = AvailableJavaHomes.getJvmInstallationMetadata(Jvm.current())
+        JvmInstallationMetadata jdkMetadata2 = AvailableJavaHomes.getJvmInstallationMetadata(AvailableJavaHomes.differentJdk)
+
+        buildFile << """
+            compileJava {
+                options.forkOptions.javaHome = file("${jdkMetadata2.javaHome}")
+            }
+        """
+
+        file("src/main/java/Foo.java") << """
+            public class Foo { }
+        """
+
+        def task = ":compileJava"
+
+        when:
+        runWithInstallation(jdkMetadata1, task)
+        def events = toolchainEvents(task)
+        then:
+        executedAndNotSkipped(task)
+        assertToolchainUsages(events, jdkMetadata2, "JavaCompiler")
```

Comment:
That is interesting. Why does `jdkMetadata1` not show up?

Reply:
I presume it is because of the way the toolchain selection is implemented in the `JavaCompile`:
https://github.com/gradle/gradle/blob/81fcbf4528111618cb2116f0d530f6957b030656/subprojects/language-java/src/main/java/org/gradle/api/tasks/compile/JavaCompile.java#L236-L263

It may also be due to the fact that `JavaCompile` (along with `Test` and `Javadoc`) does not set a convention on the tool property.

Reply:
Right, though there still is https://github.com/gradle/gradle/blob/b81eac42a17b7249ee7d2fe76e81e0326e0b1f94/subprojects/language-java/src/main/java/org/gradle/api/tasks/compile/JavaCompile.java#L123-L133

I suppose if you configure a toolchain in the `java` project block, then two will show up. Can you add/modify a test for this?

Reply:
If additionally a toolchain is configured via the `java` block, then a validation kicks in that fails the `compileJava` task.
```
java {
    toolchain {
        languageVersion = JavaLanguageVersion.of(${jdkMetadata1.languageVersion.majorVersion})
    }
}
```

```
Execution failed for task ':compileJava'.
> Must not use `javaHome` property on `ForkOptions` together with `javaCompiler` property
```